habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py

```diff
@@ -315,3 +320,425 @@ def get_all_object_ids(sim: habitat_sim.Simulator) -> Dict[int, str]:
+) -> List[mn.Vector3]:
+    """
+    Get a list of object keypoints in global space.
+    0th point is the center of mass (CoM), others are bounding box corners.
+
+    :param objectA: The ManagedRigidObject from which to extract keypoints.
+
+    :return: A set of global 3D keypoints for the object.
+    """
+
+    local_keypoints = [mn.Vector3(0)]
+    local_keypoints.extend(
+        get_bb_corners(objectA.root_scene_node.cumulative_bb)
+    )
+    global_keypoints = [
+        objectA.transformation.transform_point(key_point)
+        for key_point in local_keypoints
+    ]
+    return global_keypoints
+
+
+def object_keypoint_cast(
+    sim: habitat_sim.Simulator,
+    objectA: habitat_sim.physics.ManagedRigidObject,
+    direction: mn.Vector3 = None,
+) -> List[habitat_sim.physics.RaycastResults]:
+    """
+    Compute's object global keypoints, casts rays from each in the specified direction and returns the resulting RaycastResults.
+    Index 0 in the list is the CoM, others are corners.
+
+    :param sim: The Simulator instance.
+    :param objectA: The ManagedRigidObject from which to extract keypoints and raycast.
+    :param direction: Optionally provide a unit length global direction vector for the raycast. If None, default to -Y.
+
+    :return: A list of RaycastResults, one from each object keypoint.
+    """
+
+    if direction is None:
+        # default to downward raycast
+        direction = mn.Vector3(0, -1, 0)
+
+    global_keypoints = get_object_global_keypoints(objectA)
+    return [
+        sim.cast_ray(habitat_sim.geo.Ray(keypoint, direction))
+        for keypoint in global_keypoints
+    ]
+
+
+def get_object_set_from_id_set(
+    sim: habitat_sim.Simulator,
+    id_set: List[int],
+    ao_link_map: Optional[Dict[int, int]] = None,
+) -> List[
+    Union[
+        habitat_sim.physics.ManagedRigidObject,
+        habitat_sim.physics.ManagedArticulatedObject,
+    ]
+]:
+    """
+    Get the ManagedObjects from a set of object_ids.
+
+    :param sim: The Simulator instance.
+    :param id_set: The set of object ids for which ManagedObjects are desired.
+    :param ao_link_map: A pre-computed map from link object ids to their parent ArticulatedObject's object id.
+
+    :return: a list of tuples, first element is a ManagedObject, second is an optional link index.
+    """
+
+    if ao_link_map is None:
+        # Note: better to pre-compute this and pass it around
+        ao_link_map = get_ao_link_id_map(sim)
+
+    rom = sim.get_rigid_object_manager()
+    aom = sim.get_articulated_object_manager()
+    rigids = [
+        (rom.get_object_by_id(ro_id), None)
+        for ro_id in id_set
+        if rom.get_library_has_id(ro_id)
+    ]
+    aos = [
+        (
+            aom.get_object_by_id(ao_link_map[ao_id]),
+            aom.get_object_by_id(ao_link_map[ao_id]).link_object_ids[ao_id],
+        )
+        for ao_id in id_set
+        if ao_id in ao_link_map
+    ]
+
+    return rigids + aos
+
+
+def above(
```

Comment:
is it possible to create a wrapper on top of this function that takes in names of the two objects/object-receptacle and checks whether one is above/on top of another? That might be easier to use with an LLM for evaluation

Reply:
I may add a general purpose function to do this for all queries generically. Thinking about it.